File: benchmark/src/inverse.swift
import Foundation

func swapRows(matrix: inout [[Double]], row1: Int, row2: Int, n: Int) {
    let temp = matrix[row1]
    matrix[row1] = matrix[row2]
    matrix[row2] = temp
}

func gaussJordan(matrix: inout [[Double]], inverse: inout [[Double]], n: Int) {
    for i in 0..<n {
        if matrix[i][i] == 0 {
            for j in (i + 1)..<n {
                if matrix[j][i] != 0 {
                    swapRows(matrix: &matrix, row1: i, row2: j, n: n)
                    swapRows(matrix: &inverse, row1: i, row2: j, n: n)
                    break
                }
            }
        }
        let pivot = matrix[i][i]
        for j in 0..<n {
            matrix[i][j] /= pivot
            inverse[i][j] /= pivot
        }
        for j in 0..<n {
            if j != i {
                let factor = matrix[j][i]
                for k in 0..<n {
                    matrix[j][k] -= factor * matrix[i][k]
                    inverse[j][k] -= factor * inverse[i][k]
                }
            }
        }
    }
}

func main() {
    let inputFile = "input/input.txt"
    let outputFile = "output/swift.txt"
    
    var input: String
    do {
        input = try String(contentsOfFile: inputFile)
    } catch {
        print("Error reading input file!")
        return
    }
    
    let lines = input.split(separator: "\n").map { String($0) }
    let n = Int(lines[0])!
    
    var matrix = [[Double]](repeating: [Double](repeating: 0.0, count: n), count: n)
    var inverse = [[Double]](repeating: [Double](repeating: 0.0, count: n), count: n)
    
    for i in 0..<n {
        let rowValues = lines[i + 1].split(separator: " ").map { Double($0)! }
        for j in 0..<n {
            matrix[i][j] = rowValues[j]
            inverse[i][j] = (i == j) ? 1.0 : 0.0
        }
    }
    
    gaussJordan(matrix: &matrix, inverse: &inverse, n: n)

    var output = ""
    for row in inverse {
        output += row.map { String($0) }.joined(separator: " ") + "\n"
    }

    do {
        try output.write(toFile: outputFile, atomically: true, encoding: .utf8)
    } catch {
        print("Error writing output file!")
    }
}

main()

File: benchmark/src/inverse.jl
function swap_rows!(matrix, row1, row2)
    temp = matrix[row1, :]
    matrix[row1, :] = matrix[row2, :]
    matrix[row2, :] = temp
end

function gauss_jordan(matrix, inverse)
    n = size(matrix, 1)
    for i in 1:n
        if matrix[i, i] == 0
            for j in i+1:n
                if matrix[j, i] != 0
                    swap_rows!(matrix, i, j)
                    swap_rows!(inverse, i, j)
                    break
                end
            end
        end
        pivot = matrix[i, i]
        for j in 1:n
            matrix[i, j] /= pivot
            inverse[i, j] /= pivot
        end
        for j in 1:n
            if j != i
                factor = matrix[j, i]
                for k in 1:n
                    matrix[j, k] -= factor * matrix[i, k]
                    inverse[j, k] -= factor * inverse[i, k]
                end
            end
        end
    end
end

function main()
    inputFile = open("input/input.txt", "r")
    outputFile = open("output/julia.txt", "w")

    n = parse(Int, readline(inputFile))

    matrix = zeros(Float64, n, n)
    inverse = zeros(Float64, n, n)
    for i in 1:n
        inverse[i, i] = 1.0
    end

    for i in 1:n
        row = split(readline(inputFile), " ")
        for j in 1:n
            matrix[i, j] = parse(Float64, row[j])
        end
    end

    gauss_jordan(matrix, inverse)

    for i in 1:n
        for j in 1:n
            write(outputFile, string(inverse[i, j]) * " ")
        end
        write(outputFile, "\n")
    end

    close(inputFile)
    close(outputFile)
end

main()
